### DISCUSSION TITLE
[Opinion/Suggestion required] Collaborative/Team using {targets}

### DISCUSSION THREAD

**Discussion by CorradoLanera** [ropensci/targets#820] [Ideas]
## Intro
I work on the same project(s) with multiple PCs (in distinct locations) and with various collaborators, involving large data files and with non-developer people requiring relevant output data in `RDS` file format (and reports).

## Two short questions
1. How can I use `{targets}` effectively (without using AWS/Gcloud, but other locally synchronized shared folders to store/load/check data and targets, e.g., OneDrive, DropBox, ...)?

2. How to share only relevant data with those who have no interest in being involved in the whole pipeline?

## My tentative answers

```r
# `.Renviron` (`.gitignore`d)

PRJ_SHARED_PATH="C:/Users/<me>/<local_synchronized_shared_service>/<proj_foder>"
INPUT_DATA_FOLDER="path/to/input/folder/inside/<proj_foder>"
OUTPUT_DATA_FOLDER="path/to/output/folder/inside/<proj_foder>/possibly/empty"
```

```r
# `.Rprofile`

targets::tar_config_set(
  store = file.path(Sys.getenv("PRJ_SHARED_PATH"), "_targets")
)
```

```r
# `_targets.R`

library(targets)
library(tarchetypes)

get_input_data_path <- function() {
  file.path(
    Sys.getenv("PRJ_SHARED_PATH"),
    Sys.getenv("INPUT_DATA_FOLDER")
  )
}

import_data <- function(.dir_path) {
  file.path(.dir_path, "db.csv") |>
    normalizePath() |>
    readr::read_csv()
}

relevant_computation <- function(db) {
  2 * length(db) + 1
}

share_objects <- function(obj_list) {
  file_name <- paste0(names(obj_list), ".rds")
  
  obj_paths <- file.path(
    Sys.getenv("PRJ_SHARED_PATH"),
    Sys.getenv("OUTPUT_DATA_FOLDER"),
    file_name
  ) |>
    normalizePath(mustWork = FALSE) |>
    set_names(names(obj_list))

  # Those must be RDS
  purrr::walk2(obj_list, obj_paths, readr::write_rds)
  obj_paths
}

list(
  tar_target(
    prjFolder,
    get_input_data_path(),
    format = "file"
  ),

  tar_target(
    importedData,
    import_data(prjFolder),
    format = "qs" # just to highlight they should not need to be RDS
  ),

  tar_target(
    relevantResults,
    relevant_computation(importedData)
  ),

  tar_render(report, "report.Rmd"),

  tar_target(
    objectToShare,
    list(
      result = relevantResults
    )
  ),

  tar_target(
    shareOutput,
    share_objects(objectToShare),
    format = "file",
    pattern = map(objectToShare)
  )
)
```

That way:
-  `_targets.yaml` is overwritten every time a project session starts, including the correct local path to the shared `_targets/` folder. So, every developer on every computer has their configuration pointing to the "same" `_targets/` folder.
-  `_targets.yaml` and `.Renviron` are `.gitignore`d so that they won't create new commits anytime someone different (or the same one on another computer) works on the project. 
- In an application, I would use a `.get_prj_shared_path` function defined in the `.Rprofile` to error handle possibly missing required environmental variables.
- The hash of targets to share as `RDS` objects is dynamically used by `shareOutput`. I.e., adding more objects in the `objectToShare` list (or changing some of them) results in uploading them only and not the whole collection again and again.
- We can use a develop-optimal file format for targets in the pipeline (e.g., `{qs}`) and a user-preferred one for the subset of targets to share (e.g., `RDS`).
- `report.rmd` is deliberately rendered locally only, and the derived `HTML` would be renamed and manually copied in the shared location at its "final" version only.
- from the developing point of view, everything relevant is still tracked and shared by git/GitHub only (i.e., the local development project folder is NOT within (a sub-folder of) the shared folder!)

## Possible issue

The target `prjFolder` is invalidated every (first) time someone works on a different computer (or is another developer). That way, all its children's targets are invalidated. BUT, once {targets} has evaluated `prjFolder`, it recognizes that it is the same folder as before, and it skips all the children's targets!

> Is there an option to directly make `tar_visnetwork` realize that a folder/file is the same and not changed even if it has a different path? That to know the "real" state of the project before `tar_make` runs (that is the very first target, so the whole pipeline results outdated from the `tar_visnetwork` point of view). Note we cannot use `tar_cue("never")`, or it would never be invalidated even if the folder content change!

## Final question
Is the depicted approach the preferred/suggested solution to enforce `{targets}` in the described environment? How/where could it be improved? Are there "obvious" issues hidden from me in using that approach that you see and which could create problematic situations?

Thank you to anyone who likes to share their thoughts.
Corrado.

**Comment by wlandau** [comment]
I looked at the pipeline, and the only issue I foresee is concurrency. If two people run the pipeline and write to the same data store at the same time, then that will corrupt the project (especially the metadata). So I recommend either a solution with separate data stores and a central pool of shared files, or establishing a policy in your team about who can run the pipeline and when.
